Struct mech/feature static condensation

KlausBSautter opened the pull request:
Hi @RiccardoRossi  with this PR we added the static condensation utility. At the moment we use it in the linear 2d and linear 3d beam, (for non-linear cases some testing is necessary). I did some testing for the linear cases and get expected results. The condensation utility is written in a generic way and can now be used for every element if wished so. 
Additionally, we got rid of the boolean in the element constructors for the linear elements and inherit them now from the non-linear elements. I did the same for the cable element which now  inherits from the truss and is not triggered by some boolean.
Also internal forces such as moments and shear forces are now printed to GP along the beam for the 3D beam, 3D linear beam and 2D linear beam. 2D nonlinear coming soon as soon as I have time. 

closes #434

Klaus

philbucher committed:
Initial Commit

philbucher committed:
Frame of condensation utility done

philbucher committed:
First draft reday, needs improvement

philbucher committed:
Utility compiles

philbucher committed:
aestetics

philbucher committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter committed:
inheritance done for 2D beam

KlausBSautter committed:
inheritance finished for 3D beam

KlausBSautter committed:
/feature-static-condensation: inherited truss nl-lin

problem: pre-stress lin truss

KlausBSautter committed:
added virtual prefixes

KlausBSautter committed:
fixed linear prestress

+ added serializer

KlausBSautter committed:
finished all inheritance

todo: add cable test & link condensation utility

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter committed:
added cableelement patch test

KlausBSautter committed:
/feature-static-condensation: static condensation in beam2D and 3D

todo: back_calculations

KlausBSautter committed:
/feature-static-condensation: internal forces

moments and forces available for 2D linear beam & verified

KlausBSautter committed:
using elemental data now

- added possibility to read in vectors to the modal part as elemental data
- internal forces for systems with hinges working now

KlausBSautter committed:
added hinge tests

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

loumalouomega commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -368,34 +364,31 @@ namespace Kratos
-		{
-			const double l = this->CalculateCurrentLength();
-			const double N = this->mNodalForces[6];
+		const double l = this->CalculateCurrentLength();
```
Here the method is still called CalculateCurrentLenght(), whic is the difference?, besides, maybe could be a good idea to have a common base class for all beams and trusses (idem for shells)

philbucher replied:
base class for shells is coming soon, working on it

loumalouomega commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -51,12 +51,10 @@ namespace Kratos
-
-		CrBeamElement3D2N(IndexType NewId, GeometryType::Pointer pGeometry,
-						bool rLinear = false);
+		CrBeamElement3D2N() {};
```
Doxygen documentation please, ideam for all headers

loumalouomega commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.hpp:
```diff
@@ -0,0 +1,85 @@
+namespace Kratos
+{
+
+	class CrBeamElementLinear3D2N : public CrBeamElement3D2N
```
Ok, I see now there is a base class, Ok, could be something more general for beams and trusses?

loumalouomega commented on applications/StructuralMechanicsApplication/custom_utilities/static_condensation_utility.cpp:
```diff
@@ -0,0 +1,240 @@
+            KRATOS_CATCH("") 
+        }
+
+		std::vector<int> CreateRemainingDofList(
```
Tabulatiosn and spaces, be aware

KlausBSautter replied:
that is strange. In my editor everything is fine but here it looks like that....

loumalouomega commented on applications/StructuralMechanicsApplication/custom_utilities/static_condensation_utility.h:
```diff
@@ -0,0 +1,74 @@
+namespace Kratos
+{
+	namespace StaticCondensationUtility
+	{
```
Please document a little bit

loumalouomega reviewed (changes requested):
In general is Ok, but the lack of documentation is disturbing. Additionally please think about define common classes among beams and trusses

philbucher commented:
![240h60](https://user-images.githubusercontent.com/25484702/35829229-78144888-0ac2-11e8-83ba-7a74bb7c3ce7.jpg)

philbucher commented:
Now on a more serious note, thx for your comments @loumalouomega 

We were also thinking abt a common base class, but eventually we decided to go this way. I will discuss again with Klaus tmr, but in general there were not too many functions that beams and trusses have in common

loumalouomega commented:
You find me out, I was just asking, the real reason was the missing documentation

loumalouomega commented:
Visual Studio does that, doesn't differ between tab and spaces

El 6 feb. 2018 9:01 a. m., "Klaus Bernd Sautter" <notifications@github.com>
escribió:

> *@KlausBSautter* commented on this pull request.
> ------------------------------
>
> In applications/StructuralMechanicsApplication/custom_utilities/static_
> condensation_utility.cpp
> <https://github.com/KratosMultiphysics/Kratos/pull/1452#discussion_r166211298>
> :
>
> > +                                 remaining_dof_list, num_dofs_remaining, num_dofs_remaining);
> +
> +            FillSchurComplements(sub_matrices[1], rLeftHandSideMatrix, remaining_dof_list,
> +                                 rDofList, num_dofs_remaining, num_dofs_condensed);
> +
> +            FillSchurComplements(sub_matrices[2], rLeftHandSideMatrix, rDofList,
> +                                 remaining_dof_list, num_dofs_condensed, num_dofs_remaining);
> +
> +            FillSchurComplements(sub_matrices[3], rLeftHandSideMatrix, rDofList,
> +                                 rDofList, num_dofs_condensed, num_dofs_condensed);
> +
> +            return sub_matrices;
> +            KRATOS_CATCH("")
> +        }
> +
> +		std::vector<int> CreateRemainingDofList(
>
> that is strange. In my editor everything is fine but here it looks like
> that....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/1452#discussion_r166211298>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEMgG6sViOWWLH1K7C_4YD2Js48qWPUks5tSAbmgaJpZM4R6AkT>
> .
>

philbucher commented:
@KlausBSautter now the serialization of `bounded_vector` and `bounded_matrix` is available, can you please use it?

KlausBSautter committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into StructMech/feature-static-condensation

philbucher committed:
Fixing indentiation

philbucher commented:
@RiccardoRossi 
To use the condensation you have to set the variable `CONDENSED_DOF_LIST` as `ElementalData` in the mdpa file. E.g.
~~~
Begin ElementalData CONDENSED_DOF_LIST
  2 [1] (2,5) // this beam has moment hinges at both ends
End ElementalData
~~~

We decided against the int list you proposed (e.g. 001001) because this would become very long for elements with more dofs

If you are ok with this we would ask javi to implement it in GiD

KlausBSautter committed:
Merge branch 'StructMech/feature-static-condensation' of https://github.com/KratosMultiphysics/Kratos into StructMech/feature-static-condensation

KlausBSautter committed:
fixed serializer for beam + fixed intends

KlausBSautter committed:
added documentation to the utility header

KlausBSautter committed:
fix type

KlausBSautter commented:
@loumalouomega I added the documentation to the utility, is this ok?

RiccardoRossi commented:
@philbucher to be honest i don't love this solution, but i should say i don't have a better one.

     Begin ElementalData CONDENSED_DOF_LIST
        2 [1] (2,5) // this beam has moment hinges at both ends
     End ElementalData

as a small remark, i think that the [1] should be a [2] since it is a vector of size 2

loumalouomega commented:
Is OK, but add to the elements too if possible

KlausBSautter committed:
added documentation to element headers

KlausBSautter commented:
@loumalouomega  I added that
@RiccardoRossi  yes it should be [2]

loumalouomega commented:
You can merge in peace

philbucher commented:
Klaus can you please wait until tmr with merging?

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cable_element_3D2N.cpp:
```diff
@@ -0,0 +1,174 @@
+
+
+	void CableElement3D2N::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+										VectorType& rRightHandSideVector,
```
@KlausBSautter I think there are some tabs instead of spaces here too.
Maybe you can automatically remove all tabs and replace them with 4 spaces

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cable_element_3D2N.cpp:
```diff
@@ -0,0 +1,174 @@
+		 TransformationMatrix = ZeroMatrix(msLocalSize, msLocalSize);
+
+		this->CreateTransformationMatrix(TransformationMatrix);
+		const double InternalStrainGL = this->CalculateGreenLagrangeStrain();
```
`internal_strain_green_lagrange`

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cable_element_3D2N.cpp:
```diff
@@ -0,0 +1,174 @@
+	
+	void CableElement3D2N::save(Serializer& rSerializer) const
+	{
+		KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
```
the baseclass it the truss, not the element!

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cable_element_3D2N.hpp:
```diff
@@ -0,0 +1,91 @@
+
+// Project includes
+#include "custom_elements/truss_element_3D2N.hpp"
+#include "includes/element.h"
```
not needed, pls remove

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -203,16 +309,24 @@ namespace Kratos
 
 		int mIterationCount = 0;
-		bool mIsLinearElement = false;
 		bool mIsLumpedMassMatrix = false;
```
I think this can be removed => get it from the properties when needed

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+				Vector dof_list_input = this->GetValue(CONDENSED_DOF_LIST);
+				std::vector<int> dofList(dof_list_input.size());
+				for (size_t i=0;i<dof_list_input.size();++i) dofList[i]=dof_list_input[i];
+				StaticCondensationUtility::CondenseLeftHandSide(*this,rLeftHandSideMatrix,dofList);
```
indentiation

KlausBSautter replied:
where exactly?

philbucher replied:
Now I don't see it any more
Maybe you can use the link I sent you by mail, where one can replace tabs with spaces in vscode

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+			rOutput.resize(write_points_number);
+		}
+
+		Matrix LeftHandSideMatrix = this->CreateElementStiffnessMatrix_Total();
```
`LHS_matrix` or `left_hand_side_matrix`

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+
+		Matrix LeftHandSideMatrix = this->CreateElementStiffnessMatrix_Total();
+
+		Vector NodalDeformation = ZeroVector(msElementSize);
```
naming

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+		}
+		//// end static back condensation
+
+		Vector Stress = prod(LeftHandSideMatrix, NodalDeformation); 
```
lowercase

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+
+	void CrBeamElementLinear2D2N::save(Serializer& rSerializer) const
+	{
+		KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
```
Element is not the baseclass!

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.hpp:
```diff
@@ -0,0 +1,110 @@
+
+// Project includes
+#include "custom_elements/cr_beam_element_2D2N.hpp"
+#include "includes/element.h"
```
not needed, pls remove

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,272 @@
+			rOutput.resize(write_points_number);
+		}
+
+		Matrix LeftHandSideMatrix = CreateElementStiffnessMatrix_Material();
```
same as above

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,272 @@
+		Matrix LeftHandSideMatrix = CreateElementStiffnessMatrix_Material();
+
+
+		Vector NodalDeformation = ZeroVector(msElementSize);
```
see above

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,272 @@
+
+
+
+		Vector Stress = prod(LeftHandSideMatrix, NodalDeformation); 
```
lower_case

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,272 @@
+
+	void CrBeamElementLinear3D2N::save(Serializer& rSerializer) const
+	{
+		KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
```
base class is not Element!

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.hpp:
```diff
@@ -0,0 +1,95 @@
+
+// Project includes
+#include "custom_elements/cr_beam_element_3D2N.hpp"
+#include "includes/element.h"
```
not needed, pls remove

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp:
```diff
@@ -423,69 +377,27 @@ namespace Kratos
-
-
+	
+				const double internal_strain_gl = this->CalculateGreenLagrangeStrain();
```
pls make sure to be consistent, see my comment above abt same variable

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,207 @@
+		rLeftHandSideMatrix = this->CreateElementStiffnessMatrix(rCurrentProcessInfo);
+
+
+		Vector NodalDeformation = ZeroVector(msLocalSize);
```
lowercase

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,207 @@
+
+	void TrussElementLinear3D2N::save(Serializer& rSerializer) const
+	{
+		KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
```
element is not the baseclass

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.hpp:
```diff
@@ -0,0 +1,108 @@
+
+// Project includes
+#include "custom_elements/truss_element_3D2N.hpp"
+#include "includes/element.h"
```
not needed, pls remove

philbucher reviewed (changes requested):
Can you please check the serializer functions, I am pretty sure you are missing things (esp the beam seems to have a lot of members that are not in the serializer functions)

Other than that its just minot comments

philbucher commented:
@KlausBSautter you can remove the trailing whitespaces in VSCode while saving with a setting, see [here](https://stackoverflow.com/questions/30884131/remove-trailing-spaces-automatically-or-with-a-shortcut)

KlausBSautter committed:
fixed variable naming

now hopefully fitting to the style guide

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter commented:
@philbucher  I changed variable namings for all my elements. Sry I will try to stick to the style guide for future work. :)

KlausBSautter committed:
not saving boolean for lumped mass matrix anymore

philbucher committed:
removing the now obsolete variable TRUSS_IS_CABLE

philbucher committed:
Changing LUMPED_MASS_MATRIX to USE_CONSISTENT_MASS_MATRIX

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:
```diff
@@ -279,12 +271,8 @@ namespace Kratos
-				rRightHandSideVector -= this->F_int_global;
-			}
+			rRightHandSideVector = ZeroVector(msElementSize);
+			rRightHandSideVector -= this->mInternalGlobalForces;
```
this should be noalias.

KlausBSautter replied:
ok

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:
```diff
@@ -279,12 +271,8 @@ namespace Kratos
-				rRightHandSideVector = ZeroVector(msElementSize);
-				rRightHandSideVector -= this->F_int_global;
-			}
+			rRightHandSideVector = ZeroVector(msElementSize);
```
here it will silently resize if the size does not match, but the price for it is to create a temporary. 
If you are sure the size is already correct, then do

        rRightHandSideVector.clear()

or
        
     noalias(rRightHandSideVector) = ZeroVector(msElementSize);

KlausBSautter replied:
my reaction forces are 0 if I do this

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:
```diff
@@ -310,17 +297,17 @@ namespace Kratos
-		bounded_vector<double,3> EquivalentLineLoad = ZeroVector(3); 
-		bounded_vector<double,msElementSize> BodyForcesGlobal = ZeroVector(msElementSize);
+		bounded_vector<double,3> equivalent_line_load = ZeroVector(3); 
+		bounded_vector<double,msElementSize> body_forces_global = ZeroVector(msElementSize);
```
you are completely overwriting ths vector on line 322. You don't need the zero initialization

KlausBSautter replied:
I am not writing values to the 3rd and 6th entry so I need the zero init.

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:
```diff
@@ -353,83 +340,71 @@ namespace Kratos
 		const double numerical_limit = std::numeric_limits<double>::epsilon();
-		Vector GeometricOrientation = ZeroVector(3);
-		GeometricOrientation[0] = this->GetGeometry()[1].X()
+		Vector geometric_orientation = ZeroVector(3);
```
I think there is an utility in the core that does exactly this orientation. Can u use that instead? it is just to centralize

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp:
```diff
@@ -115,52 +120,148 @@ namespace Kratos
+	private:
 
+		// stores the deformation modes
+		bounded_vector<double,msLocalSize> mDeformationForces = ZeroVector(msLocalSize);
```
why do you need to store these vectors? you could pass them as arguements to the functions as needed.

KlausBSautter replied:
I need the DeformationForces from the previous step to calculate the geometric part of my stiffness matrix, thus I need to store this variable

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -404,46 +397,46 @@ namespace Kratos
-		ReferenceCoordinates[3] = this->GetGeometry()[1].X0();
-		ReferenceCoordinates[4] = this->GetGeometry()[1].Y0();
-		ReferenceCoordinates[5] = this->GetGeometry()[1].Z0();
+		array_1d<double, msDimension> direction_vector_x = ZeroVector(msDimension);
```
i think there is an utility in the core which should centalize all of this. can u take a look?

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -454,20 +447,20 @@ namespace Kratos
-		this->mNX0 = DirectionVectorX;
-		this->mNY0 = DirectionVectorY;
-		this->mNZ0 = DirectionVectorZ;
+		this->mNX0 = direction_vector_x;
```
why are u storing both this and the quaternion?

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -622,50 +615,50 @@ namespace Kratos
-		q.RotateVector3(rotatedNX0);
-		q.RotateVector3(rotatedNY0);
-		q.RotateVector3(rotatedNZ0);
+		Vector rotated_nx0 = this->mNX0;
```
should this be &? or do you want a copy?

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -834,13 +825,13 @@ namespace Kratos
 
 		rDampingMatrix = ZeroMatrix(msElementSize, msElementSize);
```
noalias

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -862,8 +853,8 @@ namespace Kratos
 
-		rDampingMatrix += alpha * MassMatrix;
-		rDampingMatrix += beta  * StiffnessMatrix;
+		rDampingMatrix += alpha * mass_matrix;
```
noalias both lines

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -886,7 +877,7 @@ namespace Kratos
-			EquivalentLineLoad += A * rho*
+			equivalent_line_load += A * rho*
 				this->GetGeometry()[i].
 				FastGetSolutionStepValue(VOLUME_ACCELERATION)*Ncontainer(0, i);
```
noalias(equivalent_line_load) += (A * rho* Ncontainer(0, i) ) *			FastGetSolutionStepValue(VOLUME_ACCELERATION)

KlausBSautter committed:
adjusted beam element to new variable naming

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1032,27 +1014,10 @@ namespace Kratos
 	{
 		KRATOS_TRY;
 		rRightHandSideVector = ZeroVector(msElementSize);
```
noalias

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1063,10 +1028,10 @@ namespace Kratos
-		this->CalculateTransformationMatrix(TransformationMatrix);
+		bounded_matrix<double,msElementSize,msElementSize> transformation_matrix = ZeroMatrix(msElementSize);
+		this->CalculateTransformationMatrix(transformation_matrix);
 		this->mRotationMatrix = ZeroMatrix(msElementSize);
```
noalias both lines

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1080,45 +1045,31 @@ namespace Kratos
 
 		bounded_matrix<double,msElementSize,msElementSize> aux_matrix = ZeroMatrix(msElementSize);
-		aux_matrix = prod(TransformationMatrix, rLeftHandSideMatrix);
+		aux_matrix = prod(transformation_matrix, rLeftHandSideMatrix);
```
take out the zero initialization 
   
      bounded_matrix<double,msElementSize,msElementSize> aux_matrix =  prod(TransformationMatrix, rLeftHandSideMatrix);

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1080,45 +1045,31 @@ namespace Kratos
 		bounded_matrix<double,msElementSize,msElementSize> aux_matrix = ZeroMatrix(msElementSize);
-		aux_matrix = prod(TransformationMatrix, rLeftHandSideMatrix);
+		aux_matrix = prod(transformation_matrix, rLeftHandSideMatrix);
 		rLeftHandSideMatrix = prod(aux_matrix,
```
noalias

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1080,45 +1045,31 @@ namespace Kratos
+		aux_matrix = prod(transformation_matrix, rLeftHandSideMatrix);
 		rLeftHandSideMatrix = prod(aux_matrix,
-			Matrix(trans(TransformationMatrix)));
+			Matrix(trans(transformation_matrix)));
```
you can use the trans directly here, without allocating the matrix.

matrix allocation is need only in the triple product

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1174,32 +1125,32 @@ namespace Kratos
 
 		this->mTotalNodalDeformation = ZeroVector(msElementSize);
-		this->mTotalNodalDeformation = actualDeformation;
+		this->mTotalNodalDeformation = actual_deformation;
```
zero initialization is not needed since you are assigning it. also use noalias

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1217,56 +1168,41 @@ namespace Kratos
-
+		bounded_matrix<double,msElementSize,msElementSize> transformation_matrix = this->mRotationMatrix;
+		// stress = ZeroVector(msElementSize);
+		Vector stress = this->mNodalForces;
```
you are overwriting it in the following line (and also use noalias)

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -1741,34 +1667,34 @@ namespace Kratos
 		}
 		this->mRotationMatrix = ZeroMatrix(msDimension);
-		this->mRotationMatrix = RotationMatrix;
+		this->mRotationMatrix = rotation_matrix;
```
noalias. Also the previous line is doing nothing

KlausBSautter committed:
changed boolean naming

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -26,10 +26,17 @@
-	private:
+	protected:
 		//const values
 		static constexpr int msNumberOfNodes = 2;
```
this is OK,
     just as a comment, you could get it as
     const int Dim = GetGeometty().WorkingSpaceDimension()

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -77,19 +82,54 @@ namespace Kratos
+		/**
+         * @brief This function calculates the elastic part of the total stiffness matrix
+         */
 		bounded_matrix<double,msElementSize,msElementSize> CreateElementStiffnessMatrix_Material();
```
Minor comment: while it is nice to have a function to return something, take care this is kind of risky from a performance point of view. You are relying on the optimizer to take out the temporary (however most of todays optimizers should do this ok)

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -178,23 +271,36 @@ namespace Kratos
 		Vector mTotalNodalDeformation;
 		Vector mIncrementDeformation;
-		Matrix mLHS, mRotationMatrix;
+		Matrix mRotationMatrix;
```
you shouldn't need to store the rotation matrix. you already have the quaternions

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -178,23 +271,36 @@ namespace Kratos
 
 	private:
 		Vector mNX, mNY, mNZ;
 		Vector mTotalNodalDeformation;
```
what is this needed for? as a minimum comment here what it should cotain.
also ideally do not store this but recompute it as needed (if that is possible)

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp:
```diff
@@ -178,23 +271,36 @@ namespace Kratos
 		Vector mIncrementDeformation;
-		Matrix mLHS, mRotationMatrix;
+		Matrix mRotationMatrix;
 		bounded_matrix<double,msElementSize,msElementSize> mRotationMatrix0;
```
this rotation 0 should be easily achievable if you have the orientation vectors

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_2D2N.cpp:
```diff
@@ -0,0 +1,249 @@
+
+		bounded_matrix<double,msElementSize,msElementSize> transformation_matrix = this->CreateRotationMatrix();
+		// calculate local displacements
+		nodal_deformation = prod(Matrix(trans(transformation_matrix)),nodal_deformation);
```
use noalias. also you don't need to define the temporary here

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,272 @@
+		bounded_matrix<double,msElementSize,msElementSize>
+		 transformation_matrix = this->GetReferenceRotationMatrix();
+		rLeftHandSideMatrix = ZeroMatrix(msElementSize, msElementSize);
+		rLeftHandSideMatrix +=
```
noalias

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,207 @@
+
+		KRATOS_TRY
+		//resizing the matrices + create memory for LHS
+		rLeftHandSideMatrix = ZeroMatrix(msLocalSize, msLocalSize);
```
take out this, and put a noalias in the next line

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.cpp:
```diff
@@ -0,0 +1,207 @@
+		Vector nodal_deformation = ZeroVector(msLocalSize);
+		this->GetValuesVector(nodal_deformation, 0);
+		rRightHandSideVector = ZeroVector(msLocalSize);
+		rRightHandSideVector -= prod(rLeftHandSideMatrix, nodal_deformation);
```
noalias(rRightHandSideVector) = - ... (and take out the zero initialization)

RiccardoRossi reviewed (changes requested):
hi @KlausBSautter you are still missing very many opportunities for optimization. in particilar take care of the noalias. 

also please make an effort to reduce what you store as member variable. I am pretty sure you can almost reduce it to zero...

philbucher reviewed (dismissed):
ok from my side now

philbucher commented:
@KlausBSautter also take a look at the codacy suggestions

KlausBSautter committed:
added noalias

+ one codacy fix

KlausBSautter committed:
Merge remote-tracking branch 'origin/StructMech/feature-static-condensation' into StructMech/feature-static-condensation

KlausBSautter committed:
more fixing

next: 3dbeam

KlausBSautter committed:
worked on 3dbeam

KlausBSautter committed:
added internal force ouput for nonlinear 2d beam

+ more noalias

KlausBSautter commented:
@RiccardoRossi I added noaliases. But sometimes using noalias leads to wrong results so I did not use it everywhere (see the code). Also I changed a lot of the 3d beam code. everything which is left is necessary as class members. 
Additionally the output for internal forces+moments is now available for 2d and 3d beam non-linear as well as linear

loumalouomega commented:
noalias implies that the operation does not use the destination vector/matrix (except += or -=). but for example cannot be used for:

~~~cpp
noalias(x) = prod(A,x);
~~~

KlausBSautter committed:
noalias added

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter commented:
@loumalouomega  thanks. I was confused that it works for += & -= etc... but now it should be fine 👍

KlausBSautter committed:
more improvements

pooyan-dadvand commented:
Nice work! :+1: But the PR title does nothing to do with adding so many new elements. 

I should add that I don't like at all the input:

```
Begin ElementalData CONDENSED_DOF_LIST
  2 [1] (2,5) // this beam has moment hinges at both ends
End ElementalData
```
From user side `CONDENSED_DOF_LIST` say nothing about adding a hinge or sliding end to the element. 

The Dof numbering is different in each element and this may lead to sever confusion in writing input files.  

Unfortunately like @RiccardoRossi I don't have good solution but we may improve this one by defining a mask rule (like old SAP90) which gives the fixity of each node of element: 

``` 
Begin ElementalData CONDENSED_DOF_LIST
element_id  [number_nodes_per_element](d1,d2,d3,r1,r2,r3, ....) 
2 [2](1,1,1,1,1,0, 1,1,1,1,1,0)// this beam has moment hinges at both ends
End ElementalData
```
I know that this is way more verbose but at least can be defined in a unique way....

philbucher commented:
@pooyan-dadvand no new elements are added in terms of functionalities
Only the existing ones were splitted up in a logical way (e.g. the differentiation btw linear and non-linear element was previously done through a flag, now its a derived element that overrides the corresponding functions)

philbucher commented:
@roigcarlo Is there a way to make the appveyor print fewer warnings abt unknown compiler versions ?
It is quite hard to read

roigcarlo commented:
@philbucher According to riccardo there should be a macro that we can define, but I couldn't make it work with appveyor version of boost. Anyway, in the top of the log there is a "messages" tab you can inspect. Errors and warning should be grouped there.

roigcarlo commented:
Also I think @rubenzorrilla is fixing that one. I happened because we merged a branch that was created before appveyor was running so it bypassed the check

philbucher commented:
cool thx

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter committed:
codacy fix

KlausBSautter commented:
@pooyan-dadvand  I know the way we define it at the moment is far from optimal.
But with your idea, I always have to give a vector with a lot of entries. 
So also @RiccardoRossi  what do you want? If the current solution is ok for the moment it would be nice to merge this branch

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

philbucher commented:
ping @RiccardoRossi can we merge this? 
It would be good to be able to test it in master before release to make sure everything works as it should.

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -99,14 +99,270 @@ namespace Kratos
+		{
+			int index = i * msDimension * 2;
+
+			rValues[index] = this->GetGeometry()[i]
```
...here you would have best done

      const auto& acc = his->GetGeometry()[i].FastGetSolutionStepValue(ACCELERATION,step)
      rvalues[index] = acc[0];
      rvalues[index+1] = acc[1];
      rvalues[index+2] = acc[2];

same for angular acceleration

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -99,14 +99,270 @@ namespace Kratos
+		for (int i = 0; i < msNumberOfNodes; ++i)
+		{
+			int index = i * msDimension * 2;
+			rValues[index] = this->GetGeometry()[i].
```
same for velocity and angular_velocity

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -99,14 +99,270 @@ namespace Kratos
+				.GetSolutionStepValue(DISPLACEMENT_X, Step);
+			rValues[index + 1] = this->GetGeometry()[i]
+				.GetSolutionStepValue(DISPLACEMENT_Y, Step);
+			rValues[index + 2] = this->GetGeometry()[i]
```
same for displacement and rotation

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -444,81 +687,91 @@ namespace Kratos
 		//update local CS
 		Matrix aux_rotation_matrix = ZeroMatrix(msDimension);
-		aux_rotation_matrix = this->UpdateRotationMatrixLocal();
+		aux_rotation_matrix = this->UpdateRotationMatrixLocal(Bisectrix,VectorDifference);
```
Matrix aux_rotation_matrix =  this->UpdateRotationMatrixLocal(Bisectrix,VectorDifference);

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -605,13 +850,13 @@ namespace Kratos
-		vector_diff = this->mQuaternionSCA_A * this->mQuaternionVEC_B;
-		vector_diff -= this->mQuaternionSCA_B * this->mQuaternionVEC_A;
-		vector_diff += MathUtils<double>::CrossProduct(this->mQuaternionVEC_A,
+		VectorDifference = ZeroVector(msDimension);
```
this initialization is for nothing, since you overwrite it in the next line

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -621,364 +866,157 @@ namespace Kratos
+
+		bounded_matrix<double,msDimension,msDimension> n_xyz = ZeroMatrix(msDimension);
+		for (unsigned int i = 0; i < msDimension; ++i) {
+			n_xyz(i, 0) = -1.0 * rotated_coordinate_system(i, 0);
```
take out the multiplication by 1.0. use instead

        n_xyz(i, 0) = - rotated_coordinate_system(i, 0);
	n_xyz(i, 1) = rotated_coordinate_system(i, 1);
        ....

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp:
```diff
@@ -621,364 +866,157 @@ namespace Kratos
+		if(this->mIterationCount!=0)
+		{Vector rotated_nx0 = ZeroVector(msDimension);
+		for (unsigned int i = 0; i < msDimension; ++i) rotated_nx0[i] = this->mLocalRotationMatrix(i, 0);
+		Vector temp_vector = ZeroVector(msDimension);
```
what do u need this zero initialization for?

KlausBSautter replied:
as far as I understand this is the vector which will contain the result of the crossproduct calculation

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -79,10 +79,10 @@ namespace Kratos
 		Vector nodal_deformation = ZeroVector(msElementSize);
```
you don't need to initialize to zero, simply to have it of the correct size

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_linear_3D2N.cpp:
```diff
@@ -79,10 +79,10 @@ namespace Kratos
 		this->GetValuesVector(nodal_deformation);
 		rRightHandSideVector = ZeroVector(msElementSize);
-		rRightHandSideVector -= prod(left_hand_side_matrix, nodal_deformation);
+		noalias(rRightHandSideVector) -= prod(left_hand_side_matrix, nodal_deformation);
```
just to observe that you could have avoided the initialization to zero by just doing

     noalias(rRightHandSideVector) =  -prod(left_hand_side_matrix, nodal_deformation);

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/truss_element_linear_3D2N.cpp:
```diff
@@ -72,8 +72,8 @@ namespace Kratos
 		Vector nodal_deformation = ZeroVector(msLocalSize);
 		this->GetValuesVector(nodal_deformation, 0);
 		rRightHandSideVector = ZeroVector(msLocalSize);
```
you could have easily avoided this initialization to zero

RiccardoRossi reviewed (dismissed):
Approving since i see no errors and it is good to have this merged before release, however there are still quite a lot of performance pitfalls which should be hopefully fixed at a later time

RiccardoRossi commented:
guys, please speak with @jginternational so that

1 - condensation is allowed in the interface
2 - output on gauss points is visiblefrom the interface

KlausBSautter commented:
+ cable is now an element not triggered by boolean ---> important for interface with GID!!

KlausBSautter committed:
Merge branch 'master' into StructMech/feature-static-condensation

KlausBSautter committed:
using auto type

KlausBSautter committed:
using clang format

KlausBSautter commented:
@jginternational  how do we proceed with the gid interface?

philbucher commented:
@jginternational the issue [exists already](https://github.com/KratosMultiphysics/GiDInterface/issues/244)
